### PR TITLE
Generalize Make system

### PR DESCRIPTION
**Pull request by nekenyu**

### CODE REVIEW

File: Make/Makefile.incl [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+all: $(EXE)
+
+# Build executable
+$(EXE): $(OBJECTS)
+	$(CXX) -o $@ $^ $(LIBS) $(CXXFLAGS)
+
+#
+# Library Generation
+#
+else ifdef LIB
```
Comment (nekenyu) [inline Make/Makefile.incl:68]: Note: This is not tested.

File: Makefile [hunk 2]
```diff
@@ -1,37 +1,13 @@
 all:
 
-PROG = bill
+compiler := src/compiler
```
Comment (nekenyu) [inline Makefile:3]: Independent definitions where each Makefile is allows for manual directory dependency.

File: Makefile [hunk 3]
```diff
@@ -1,37 +1,13 @@
...
-#CXXFLAGS+= -lboost_unit_test_framework		# these lines required for test
-#LIBS+= boost/test/included/unit_test.hpp
-#LIBS+= boost_unit_test
+DIRS+=$(compiler)
 
-#unit_test.o: unit_test.hpp
-#	g++ -o unit_test.o -c boost/test/unit_test.hpp ${CXXFLAGS}
+TARGETS = all clean cleaner cleanest
+.PHONY: $(TARGETS) $(DIRS)
+$(TARGETS): $(DIRS)
```
Comment (nekenyu) [inline Makefile:9]: Make magic.  Every valid target depends on all the directories to be (sub)make-d.  We then give the actual command goals in the target to build each directory.

File: src/compiler/Makefile [hunk 4]
```diff
@@ -0,0 +1,5 @@
+all:
```
Comment (nekenyu) [inline src/compiler/Makefile:1]: I know, its a paint to create a Makefile for every executable and library, but...  You have to get the config right and doing it in the top-level will drive us all crazy.  So, we've made it so you specify what you are making (an EXE or LIB), set a few options, then include the generic Makefile.incl. Pretty simple, right?

File: Make/Makefile.incl [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+CXXFLAGS+= -O3
+
+# Also generate dependency files when compiling objects
+CXXFLAGS_MAKE_DEPEND=-MMD
+
+# Archiver Info
+AR=ar
+ARFLAGS+=rcs
+
+# File definitions
```
Comment (nekenyu) [inline Make/Makefile.incl:31]: Do all this dynamically.  Doesn't work from Windows, sorry to say ;-)  We'll deal with that later.

**Comment by jsmith1024** [comment]
Quwstion:
Why does it appear you removed the components required for boost test?

**Comment by nekenyu** [comment]
They were in the wrong place and not working.  They would be added back when a change is made to support unit test.

**Review by jsmith1024 — APPROVED** [review]

**Pull request merged**